gateway/ln-gateway/src/lib.rs

```diff
@@ -551,8 +551,17 @@ impl Gateway {
                         break;
                     };
 
-                    let htlc_request = match stream.next().await {
-                        Some(Ok(htlc_request)) => htlc_request,
+                    let payment_request = tokio::select! {
+                        payment_request = stream.next() => {
+                            payment_request
+                        }
+                        () = self.is_shutting_down_safely() => {
```

Comment:
As discussed offline, I moved this to after the `select!` and after we take the lock, so that we can be sure that `GatewayState` doesn't change before `handle_htlc` is called.